File: Quadrotor/la_x_response.py
import rosbag
import numpy as np
import matplotlib.pyplot as plt
import sys

def get_data(bag_filename):
    time_command = []
    time_pose = []
    time_euler = []

    x_command = []
    y_command = []
    z_command = []

    roll = []
    pitch = []
    yaw = []

    x_pose = []
    y_pose = []
    z_pose = []

    bag = rosbag.Bag(bag_filename)
    topics = set([
        "/command/pose",
        "/ground_truth_to_tf/euler",
        "/ground_truth_to_tf/pose"
    ])

    for topic, msg, t in bag.read_messages(topics=topics):
        if topic == '/command/pose':
            time_command.append(msg.header.stamp.to_sec())
            x_command.append(msg.pose.position.x)
            y_command.append(msg.pose.position.y)
            z_command.append(msg.pose.position.z)

        if topic == '/ground_truth_to_tf/euler':
            time_euler.append(msg.header.stamp.to_sec())
            roll.append(msg.vector.x)
            pitch.append(msg.vector.y)
            yaw.append(msg.vector.z)

        if topic == '/ground_truth_to_tf/pose':
            time_pose.append(msg.header.stamp.to_sec())
            x_pose.append(msg.pose.position.x)
            y_pose.append(msg.pose.position.y)
            z_pose.append(msg.pose.position.z)
    bag.close()
    return time_command, time_pose, time_euler, x_command, y_command, \
                  z_command, roll, pitch, yaw, x_pose, y_pose, z_pose

if __name__ == '__main__':
    ref_x = 1.0
    ref_y = 0.0
    ref_z = 1.0

    bag_filename = sys.argv[1]
    time_command, time_pose, time_euler, x_command, y_command, \
                  z_command, roll, pitch, yaw, x_pose, y_pose, z_pose = get_data(bag_filename)

    time_command = np.asarray(time_command)
    x_command = np.asarray(x_command)
    y_command = np.asarray(y_command)
    z_command = np.asarray(z_command)

    time_euler = np.asarray(time_euler)
    roll = np.asarray(roll)
    pitch = np.asarray(pitch)
    yaw = np.asarray(yaw)

    time_pose = np.asarray(time_pose)
    x_pose = np.asarray(x_pose)
    y_pose = np.asarray(y_pose)
    z_pose = np.asarray(z_pose)

    step_index = np.where(x_command > (ref_x / 2.0))[0][0]
    step_time = time_command[step_index]

    time_command = time_command - step_time
    time_pose = time_pose - step_time
    time_euler = time_euler - step_time

    start_index = np.where(time_pose > 0)[0][0]
    end_index = np.where(time_pose > 8)[0][0]
    SAE = np.sum(np.abs(x_pose[start_index:end_index] - ref_x)) + \
          np.sum(np.abs(y_pose[start_index:end_index] - ref_y)) + \
          np.sum(np.abs(z_pose[start_index:end_index] - ref_z))

    print('the sum of abosulte erros is', SAE)
    print('the overshoot is', np.max(x_pose) - ref_x)
    print('the ripple of pitch angle is', np.max(pitch[end_index:]) - np.min(pitch[end_index:]))

    fig, axs = plt.subplots(2, 1)

    axs[0].plot(time_command, x_command, time_pose, x_pose, linewidth=4)
    axs[0].set_xlim(-1, 8)
    axs[0].set_ylim(-0.1, 1.3)
    axs[0].set_ylabel('x (m)', fontsize=30)
    axs[0].grid(True)
    axs[0].legend(('reference', 'response'), loc='lower right', fontsize=30)
    axs[0].set_xticklabels([])
    axs[0].tick_params(axis="y", labelsize=30)
    fig.suptitle('Step Response', fontsize=30)

    axs[1].plot(time_euler, pitch, linewidth=4)
    axs[1].set_xlim(-1, 8)
    axs[1].set_ylim(-0.3, 0.4)
    axs[1].set_xlabel('time (second)', fontsize=30)
    axs[1].tick_params(axis="x", labelsize=30)
    axs[1].tick_params(axis="y", labelsize=30)
    axs[1].set_ylabel('pitch (rad)', fontsize=30)
    axs[1].grid(True)
    # fig.tight_layout()
    plt.show()

File: PID/Quadrotor/la_x_response.m
bag = rosbag('la_s_x_2019-10-29-15-03-06.bag');
bSel_command_pose = select(bag,'Topic','/command/pose');
bSel_euler = select(bag,'Topic','/ground_truth_to_tf/euler');
bSel_pose = select(bag,'Topic','/ground_truth_to_tf/pose');

msgStructs_command_pose = readMessages(bSel_command_pose,'DataFormat','struct');
msgStructs_euler = readMessages(bSel_euler,'DataFormat','struct');
msgStructs_pose = readMessages(bSel_pose,'DataFormat','struct');

Sec_command=  cellfun(@(m) double(m.Header.Stamp.Sec),msgStructs_command_pose);
Nsec_command=  cellfun(@(m) double(m.Header.Stamp.Nsec),msgStructs_command_pose);
time_command = Sec_command + Nsec_command/10^9;

Sec_pose=  cellfun(@(m) double(m.Header.Stamp.Sec),msgStructs_pose);
Nsec_pose=  cellfun(@(m) double(m.Header.Stamp.Nsec),msgStructs_pose);
time_pose = Sec_pose + Nsec_pose/10^9;

Sec_euler=  cellfun(@(m) double(m.Header.Stamp.Sec),msgStructs_euler);
Nsec_euler=  cellfun(@(m) double(m.Header.Stamp.Nsec),msgStructs_euler);
time_euler = Sec_euler + Nsec_euler/10^9;

x_command=  cellfun(@(m) double(m.Pose.Position.X),msgStructs_command_pose);
y_command=  cellfun(@(m) double(m.Pose.Position.Y),msgStructs_command_pose);
z_command=  cellfun(@(m) double(m.Pose.Position.Z),msgStructs_command_pose);

x_pose=  cellfun(@(m) double(m.Pose.Position.X),msgStructs_pose);
y_pose=  cellfun(@(m) double(m.Pose.Position.Y),msgStructs_pose);
z_pose=  cellfun(@(m) double(m.Pose.Position.Z),msgStructs_pose);

roll = cellfun(@(m) double(m.Vector.X),msgStructs_euler);
pitch = cellfun(@(m) double(m.Vector.Y),msgStructs_euler);
yaw = cellfun(@(m) double(m.Vector.Z),msgStructs_euler);

step_indexs = x_command > 0.5;
[~,step_index] = max(step_indexs);
step_time = time_command(step_index);

time_command = time_command - step_time;
time_pose = time_pose - step_time;
time_euler = time_euler - step_time;

subplot(2,1,1);
plot(time_command,x_command,'LineWidth',4);
hold on;
plot(time_pose, x_pose,'LineWidth',4);
hold on;

title('Step Response','Interpreter','latex')
ylabel({'$x (m)$'},'Interpreter','latex','Rotation',0,'HorizontalAlignment','Right');
ax = gca;
ax.XLim = [-1,8];
ax.YLim = [-0.1,1.3];
ax.TickLabelInterpreter = 'latex';
legend('reference','response','Interpreter','latex');
set(gca,'xtick',[],'xticklabel',[],'FontSize',30,'linewidth',2);

subplot(2,1,2);
plot(time_euler,pitch,'LineWidth',4);
ylabel({'$pitch (rad)$'},'Interpreter','latex','Rotation',0,'HorizontalAlignment','Right');
xlabel({'$time\, (second)$'},'Interpreter','latex');
ax = gca;
ax.TickLabelInterpreter = 'latex';
ax.XLim = [-1,8];
ax.YLim = [-0.3,0.4];
set(gca,'FontSize',30,'linewidth',2);